src/lib/updateContractInfo.ts

```diff
@@ -0,0 +1,191 @@
+//     `SELECT "blockNumber" FROM "TokenTransfer" WHERE "tokenAddress" = '${tokenAddress}' ORDER BY "blockNumber" DESC LIMIT 1;`,
+//     { raw: true, type: QueryTypes.SELECT, plain: true }
+//   )
+//   return row?.blockNumber
+// }
+
+export async function updateExistingContractInfoByBlock({
+  blockNumber,
+  goal
+}: IContractInfoJobDetailsByBlock): Promise<void> {
+  if (goal === ProcessingGoal.BACKFILL) {
+    return
+  }
+  const client = await getEthClient()
+  const blockWithTransactions = await client.getBlockWithTransactions(
+    blockNumber
+  )
+
+  const possibleContractddresses = blockWithTransactions.transactions.reduce(
+    (map: { toAddress: Set<string>; fromAddress: Set<string> }, tx) => {
+      if (tx.from) {
+        map.fromAddress.add(tx.from.toLowerCase())
+      }
+      if (tx.to) {
+        map.toAddress.add(tx.to.toLowerCase())
+      }
+
+      return map
+    },
+    { toAddress: new Set<string>(), fromAddress: new Set<string>() }
+  )
+
+  const possibleContractAddressUpdates = await ContractInfo.findAll({
+    where: {
+      address: {
+        [Op.in]: [
+          ...possibleContractddresses.fromAddress,
+          ...possibleContractddresses.toAddress
+        ]
+      }
+    }
+  })
+
+  await Bluebird.Promise.map(
+    possibleContractAddressUpdates,
+    contractInfo =>
+      updateContractInfoByTokenAddress(
+        {
+          tokenAddress: contractInfo.address,
+          transferBlockNumber: blockNumber
+        },
+        blockWithTransactions
+      ),
+    { concurrency: 5 }
+  )
+}
+
+/**
+ * Update, or add, a contracts's metadata into the `ContractInfo` table.
+ *
+ * @remarks
+ * This function not only gets metadata from the contract in the blockchain itself.
+ *
+ * @param {tokenAddress} tokenAddress - The address of the token.
+ * @param {blockNumber} blockWithTransactions - The block (with transactions) when updating contractinfo by block.
+
+ * @returns {Promise<void>} This function does not return any useful value.
+ */
+export async function updateContractInfoByTokenAddress(
+  {
+    tokenAddress,
+    transferBlockNumber,
+    goal,
+    fullUpdate = false
+  }: IContractInfoJobDetailsByTokenAddress,
+  blockWithTransactions?: BlockWithTransactions
+): Promise<void> {
+  const address = tokenAddress.toLowerCase()
+  const startTime = Date.now()
+  stats.increment('update_contract_called')
+
+  try {
+    const [contractInfo, created] = await ContractInfo.findOrCreate({
+      where: {
+        address
+      }
+    })
+
+    if (!created && goal === ProcessingGoal.BACKFILL) {
+      stats.increment('update_token_backfill')
+      return
+    }
+
+    const updateSpec =
+      fullUpdate ||
+      created ||
+      !contractInfo.updatedMetaInfoAt ||
+      differenceInDays(Date.now(), contractInfo.updatedMetaInfoAt) >=
+        CONTRACT_INFO_MAX_AGE_IN_DAYS
+
+    const updateBlockMetrics =
+      !!transferBlockNumber &&
```

Comment:
Is there a scenario where `transferBlockNumber` is not present here?

Reply:
yes. when doing enrichment refresh from the querying server. in which we just trigger refresh and it will only update the meta info (name, symbol, etc)

Reply:
gotcha!